src/GafferImage/CollectImages.cpp

```diff
@@ -0,0 +1,447 @@
+	{
+		return outPlug()->metadataPlug()->defaultValue();
+	}
+}
+
+
+
+void CollectImages::hashDataWindow( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashDataWindow( output, context, h );
+
+	const std::string layerContextVariable = layerContextVariablePlug()->getValue();
+
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+	const vector<string> &layerNames = layerNamesData->readable();
+
+	Context::EditableScope editScope( context );
+	for( unsigned int i = 0; i < layerNames.size(); i++ )
+	{
+		editScope.set( layerContextVariable, layerNames[i] );
+		inPlug()->dataWindowPlug()->hash( h );
+	}
+}
+
+Imath::Box2i CollectImages::computeDataWindow( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	Imath::Box2i dataWindow;
+
+	const std::string layerContextVariable = layerContextVariablePlug()->getValue();
+
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+	const vector<string> &layerNames = layerNamesData->readable();
+
+	Context::EditableScope editScope( context );
+	for( unsigned int i = 0; i < layerNames.size(); i++ )
+	{
+		editScope.set( layerContextVariable, layerNames[i] );
+		dataWindow.extendBy( inPlug()->dataWindowPlug()->getValue() );
+	}
+
+	return dataWindow;
+}
+
+void CollectImages::hashChannelNames( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashChannelNames( output, context, h );
+
+	const std::string layerContextVariable = layerContextVariablePlug()->getValue();
+
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+	const vector<string> &layerNames = layerNamesData->readable();
+
+	h.append( &(layerNames[0]), layerNames.size() );
+
+	Context::EditableScope editScope( context );
+	for( unsigned int i = 0; i < layerNames.size(); i++ )
+	{
+		editScope.set( layerContextVariable, layerNames[i] );
+		inPlug()->channelNamesPlug()->hash( h );
+	}
+}
+
+IECore::ConstStringVectorDataPtr CollectImages::computeChannelNames( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	StringVectorDataPtr channelNamesData = new StringVectorData();
+	std::vector<string> &outChannelNames = channelNamesData->writable();
+
+	const std::string layerContextVariable = layerContextVariablePlug()->getValue();
+
+	ConstStringVectorDataPtr layerNamesData = layerNamesPlug()->getValue();
+	const vector<string> &layerNames = layerNamesData->readable();
+
+	Context::EditableScope editScope( context );
+	for( unsigned int i = 0; i < layerNames.size(); i++ )
+	{
+		editScope.set( layerContextVariable, layerNames[i] );
+		ConstStringVectorDataPtr layerChannelsData = inPlug()->channelNamesPlug()->getValue();
+		const std::vector<string> &layerChannels = layerChannelsData->readable();
+		
+		for( unsigned int j = 0; j < layerChannels.size(); j++ )
+		{
+			std::string curName;
+			if( layerNames[i].size() > 0 )
```

Comment:
Hmm, that's a reasonable point. I still think something is wrong if we're needing to do manual string formatting in the nodes themselves though. Maybe we just clarify that the second argument to `channelName()` doesn't have to be a pure baseName?